File: examples/authentication/cc/rt/naeem/hottentot/runtime/proxy/response_reader.cc
#include "response_reader.h"
#include "tcp_client.h"

#include "../protocol_v1.h"
#include "../response.h"


namespace naeem {
  namespace hottentot {
    namespace runtime {
      namespace proxy {
        Response*
        ResponseReader::ReadResponse() {
          // Read response byte array
          uint32_t responseSerializedDataLength = 0;
          // TODO(kamran): Read operation should be buffered.
          unsigned char *responseSerializedData = tcpClient_->Read(&responseSerializedDataLength);
          // Deserialize request according to HTNP
          ::naeem::hottentot::runtime::Protocol *protocol = 
            new ::naeem::hottentot::runtime::ProtocolV1(); // TODO(kamran): Use factory.        
          ::naeem::hottentot::runtime::Response *response = protocol->DeserializeResponse(responseSerializedData, 
                                                                                          responseSerializedDataLength);
          delete responseSerializedData;
          delete protocol; // WARN(kamran): Don't delete if it is a shared pointer and the method is not the owner.
        }
      }
    }
  }
}

File: examples/authentication/java/src/main/java/runtime/ResponseReader.java
package runtime;


import runtime.exception.TcpClientReadException;

public class ResponseReader {

    private TcpClient tcpClient;
    public ResponseReader(TcpClient tcpClient) {
        this.tcpClient = tcpClient;
    }

    public Response read() throws TcpClientReadException {
        Protocol protocol = new ProtocolV1();
        byte[] serializedResponse;
        try {
            //System.out.println("-----" + tcpClient);
            serializedResponse = tcpClient.read();

        } catch (TcpClientReadException e) {
            throw e;
        }
        //System.out.println(serializedResponse);
        return protocol.deserializeResponse(serializedResponse);
    }
}
